hub/api/pytorch.py

```diff
@@ -0,0 +1,241 @@
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.unregister(self, name, rtype)
+
+    resource_tracker.register = fix_register
+    resource_tracker.unregister = fix_unregister
+    if "shared_memory" in resource_tracker._CLEANUP_FUNCS:
+        del resource_tracker._CLEANUP_FUNCS["shared_memory"]
+
+
+def _to_pytorch(dataset, transform=None, workers=1):
+    try:
+        import torch
+    except ModuleNotFoundError:
+        raise ModuleNotInstalledException("torch")
+    global torch
+    global _hub_storage_provider  # global to pass to processes, as not possible to serialize it and send as argument
+
+    # TODO, apparently boto3.client isn't safe for multiprocessing https://github.com/boto/boto3/pull/2848/files
+    # could it be working here as we're only reading data?
+    _hub_storage_provider = dataset.provider
+    return TorchDataset(dataset, transform, workers)
+
+
+class TorchDataset:
+    def __init__(self, dataset, transform=None, workers=1):
+        self.dataset = dataset  # TODO disable the memory cache
+        self.transform = transform
+        self.workers = workers
+        self.storage = self.dataset.provider
+        self.map = ProcessPool(nodes=workers).map
+
+        # contains index_map for each Tensor
+        self.all_index_maps = self.load_index_maps()
+
+        # contains meta for each Tensor
+        self.all_meta = self.load_meta()
+
+        # corresponding to each Tensor, stores index-value map. here value is the actual array at the index for the key
+        # this essentially acts as our in memory prefetch cache
+        self.all_index_value_maps = defaultdict(dict)
+
+        # for each Tensor, stores the last index that was prefetched in the prefetch cache
+        self.last_index_map = {}
+
+        # list of all the final samples generated after prefetching and transforming, currently present in memory
+        self.processed_samples = None
+        self.processed_range = slice(-1, -1)  # start and end range of processed_samples
+
+        # keeps track of names of all chunks across tensors whose data is currently prefetched
+        self.all_chunk_sets = {}
+
+        # keeps a pointer to all shared memory objects so that they don't get dereferenced between calls to getitem
+        self.all_shared_memory = []
+
+    def __len__(self):
+        return len(self.dataset)
+
+    def __getitem__(self, index):
+        for key in self.dataset.tensors:
+            # if cache hit
+            if index in self.all_index_value_maps[key]:
+                continue
+
+            # index is outside prefetch cache, clear it (and fetch more later)
+            if index != 0 and index == self.last_index_map[key] + 1:
+                del self.all_index_value_maps[key]
+                chunk_set = self.all_chunk_sets[key]
+                shared_memory_clear(chunk_set)
+
+            chunk_set = self.get_chunk_names(index, key)
```

Comment:
nit-pick: `chunk_set` should be renamed to `chunk_names` (also in the `get_chunk_names(...)` function